powertools-parameters/src/main/java/software/amazon/lambda/powertools/parameters/AppConfigProvider.java

```diff
@@ -0,0 +1,226 @@
+                response.configuration().asUtf8String() : // if we have a new value, use it
+                    establishedSession != null?
+                            establishedSession.lastConfigurationValue : // if we don't but we have a previous value, use that
+                            null; // otherwise we've got no value
+
+        // Update the cache so we can get the next value later
+        establishedSessions.put(key, new EstablishedSession(nextSessionToken, value));
+
+        return value;
+    }
+
+    @Override
+    protected Map<String, String> getMultipleValues(String path) {
+        // Retrieving multiple values is not supported with the AppConfig provider.
+        throw new RuntimeException("Not implemented");
+    }
+
+    /**
+     * Create a builder that can be used to configure and create a {@link AppConfigProvider}.
+     *
+     * @return a new instance of {@link AppConfigProvider.Builder}
+     */
+    public static AppConfigProvider.Builder builder() {
+        return new AppConfigProvider.Builder();
+    }
+
+    static class Builder {
+        private AppConfigDataClient client;
+        private CacheManager cacheManager;
+        private TransformationManager transformationManager;
+        private String environment;
+        private String application;
+
+        /**
+         * Create a {@link AppConfigProvider} instance.
+         *
+         * @return a {@link AppConfigProvider}
+         */
+        public AppConfigProvider build() {
+            if (cacheManager == null) {
+                throw new IllegalStateException("No CacheManager provided; please provide one");
+            }
+            if (environment == null) {
+                throw new IllegalStateException("No environment provided; please provide one");
+            }
+            if (application == null) {
+                throw new IllegalStateException("No application provided; please provide one");
+            }
+
+            // Create a AppConfigDataClient if we haven't been given one
+            if (client == null) {
+                AppConfigDataClientBuilder ssmClientBuilder = AppConfigDataClient.builder()
+                        .httpClientBuilder(UrlConnectionHttpClient.builder())
+                        .region(Region.of(System.getenv(SdkSystemSetting.AWS_REGION.environmentVariable())));
+
+                // AWS_LAMBDA_INITIALIZATION_TYPE has two values on-demand and snap-start
+                // when using snap-start mode, the env var creds provider isn't used and causes a fatal error if set
+                // fall back to the default provider chain if the mode is anything other than on-demand.
+                String initializationType = System.getenv().get(AWS_LAMBDA_INITIALIZATION_TYPE);
+                if (initializationType  != null && initializationType.equals(LambdaConstants.ON_DEMAND)) {
+                    ssmClientBuilder.credentialsProvider(EnvironmentVariableCredentialsProvider.create());
```

Comment:
Aren't you missing something? `ssmClientBuilder.build()` is not called, client cannot be initialised...

Reply:
Fixed! And this is why we do code reviews 🤣  🤦